File: SlidePuzzle.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class SlidePuzzle {
    public static void main(String[] args) {
        final int BOARD_SIZE = 3;
        ArrayList<String> board = new ArrayList<String>();
        board.add("  ");
        for (int i = 1; i < BOARD_SIZE*BOARD_SIZE; i++) {
            board.add(String.format("%2s", Integer.toString(i)));
        }
        ArrayList<String> sorted = new ArrayList<String>(board);
        Collections.shuffle(board);
        int moves = 0;
        Scanner key = new Scanner(System.in);
        while (true) {
            System.out.println("Moves: " + moves);
            moves++;
            System.out.println("    ");
            int counter = 1;
            for (String piece : board) {
                System.out.print(piece + " ");
                if (counter % BOARD_SIZE == 0) {
                    System.out.println("    ");
                }
                counter++;
            }
            System.out.print("\nMove: ");
            String inp = key.nextLine();
            if (inp == "q") {
                key.close();
                break;
            }
            String move = String.format("%2s", inp);
            int open_space_loc = board.indexOf("  ");
            int move_loc = 0;
            try {
                move_loc = board.indexOf(move);
            } catch (Exception e) {
                continue;
            }
            if (
                move_loc + 1 == open_space_loc
                || move_loc - 1 == open_space_loc
                || move_loc + BOARD_SIZE == open_space_loc
                || move_loc - BOARD_SIZE == open_space_loc
            ){
                Collections.swap(board, open_space_loc, move_loc);
            }
            if (board == sorted) {
                key.close();
                System.out.println("You win!");
                break;
            }
            for (int i = 0; i < 20; i++) System.out.print("-");
            System.out.println();
        }
    }
}

File: slide_puzzle.py
#!/usr/bin/env python3

import curses
import random

BOARD_SIZE = 3

def main(stdscr):
    curses.use_default_colors()
    board = [str(i).rjust(2, ' ') for i in ["  ", *range(1, BOARD_SIZE * BOARD_SIZE)]]
    random.shuffle(board) # this leads to an occasional issue where the board is not solvable
    moves = 0
    stdscr.nodelay(0)
    while True:
        stdscr.clear()
        stdscr.addstr(f"Moves: {moves}\n\n    ")
        moves += 1
        for i, piece in enumerate(board, 1):
            stdscr.addstr(f"{piece} ")
            if i % BOARD_SIZE == 0:
                stdscr.addstr("\n    ")
        stdscr.addstr("\nMove: ")
        key = stdscr.getkey()
        if key == 'q':
            return 'quit'
        move = key.rjust(2, ' ')
        open_space_loc = board.index('  ')
        try:
            move_loc = board.index(move)
        except ValueError:
            continue
        if open_space_loc in (
            move_loc + 1,
            move_loc - 1,
            move_loc + BOARD_SIZE,
            move_loc - BOARD_SIZE
        ):
            board[open_space_loc], board[move_loc] = board[move_loc], board[open_space_loc]
        if board == sorted(board):
            return 'win'
        stdscr.refresh()

if __name__ == "__main__":
    if curses.wrapper(main) == 'win':
        print("You Win!")
